File: src/main/java/com/m2t/hachikodictionary/model/Word.java
package com.m2t.hachikodictionary.model;

import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.Set;

@Data
@NoArgsConstructor
@Document("words")
public class Word {

    @Id
    private String id;

    private String title;
    private String kind;
    private String difficulty = "medium";

    private Set<String> descriptions;

    private Set<String> synonyms;

    private Set<String> antonyms;

    private Set<String> sentences;

    private Set<LearnedWord> learnedWords;

    private String fileUrl;

    private LocalDateTime audioFileCreatedAt;

    public Word(String title, String kind, String difficulty, Set<String> descriptions, Set<String> synonyms, Set<String> antonyms, Set<String> sentences, Set<LearnedWord> learnedWords, String fileUrl) {
        this.title = title;
        this.kind = kind;
        this.difficulty = difficulty;
        this.descriptions = descriptions;
        this.synonyms = synonyms;
        this.antonyms = antonyms;
        this.sentences = sentences;
        this.learnedWords = learnedWords;
        this.fileUrl = fileUrl;
    }

    public Word(String id, String title, String kind, String difficulty, Set<String> descriptions, Set<String> synonyms, Set<String> antonyms, Set<String> sentences, Set<LearnedWord> learnedWords, String fileUrl, LocalDateTime audioFileCreatedAt) {
        this.id = id;
        this.title = title;
        this.kind = kind;
        this.difficulty = difficulty;
        this.descriptions = descriptions;
        this.synonyms = synonyms;
        this.antonyms = antonyms;
        this.sentences = sentences;
        this.learnedWords = learnedWords;
        this.fileUrl = fileUrl;
        this.audioFileCreatedAt = audioFileCreatedAt;
    }

    public Word(String id, String title, String kind, String difficulty, Set<String> descriptions, Set<String> synonyms, Set<String> antonyms, Set<String> sentences, String fileUrl, LocalDateTime audioFileCreatedAt) {
        this.id = id;
        this.title = title;
        this.kind = kind;
        this.difficulty = difficulty;
        this.descriptions = descriptions;
        this.synonyms = synonyms;
        this.antonyms = antonyms;
        this.sentences = sentences;
        this.fileUrl = fileUrl;
        this.audioFileCreatedAt = audioFileCreatedAt;
    }
}

File: src/main/java/com/m2t/hachikodictionary/model/Word.kt
package com.m2t.hachikodictionary.model

import jakarta.annotation.Nullable
import jakarta.persistence.*
import org.hibernate.annotations.GenericGenerator
import org.springframework.boot.context.properties.bind.DefaultValue
import java.time.LocalDate
import java.time.LocalDateTime

@Entity
@Table(name = "words", schema = "public")
data class Word(
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator") // Generate UUID
    var id: String?,
    var title: String,
    var kind: String,

    var difficulty: String = "medium",

    @ElementCollection
    @CollectionTable(name = "word_descriptions", schema = "public")
    var descriptions: Set<String>,
    @ElementCollection
    @CollectionTable(name = "word_synonyms", schema = "public")
    var synonyms: Set<String>,
    @ElementCollection
    @CollectionTable(name = "word_antonyms", schema = "public")
    var antonyms: Set<String>,
    @ElementCollection
    @CollectionTable(name = "word_sentences", schema = "public")
    var sentences: Set<String>,

    @OneToMany(mappedBy = "word", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val learnedWords: Set<LearnedWord>,

    @Nullable
    @Column(columnDefinition = "TEXT")
    var fileUrl: String?,

    var audioFileCreatedAt: LocalDateTime?
) {
    constructor(
        title: String,
        kind: String,
        descriptions: MutableSet<String>,
        synonyms: MutableSet<String>,
        antonyms: MutableSet<String>,
        sentences: MutableSet<String>,
            fileUrl: String?,
        audioFileCreatedAt: LocalDateTime?
    ) : this(
        null,
        title,
        kind,
        "medium",
        descriptions,
        synonyms,
        antonyms,
        sentences,
        emptySet(),
        fileUrl,
        audioFileCreatedAt
    )

    constructor(
        title: String,
        kind: String,
        difficulty: String,
        descriptions: MutableSet<String>,
        synonyms: MutableSet<String>,
        antonyms: MutableSet<String>,
        sentences: MutableSet<String>,
        fileUrl: String?,
        audioFileCreatedAt: LocalDateTime?
    ) : this(
        null,
        title,
        kind,
        difficulty,
        descriptions,
        synonyms,
        antonyms,
        sentences,
        emptySet(),
        fileUrl,
        audioFileCreatedAt
    )

    constructor(
        id: String,
        title: String,
        kind: String,
        difficulty: String,
        descriptions: MutableSet<String>,
        synonyms: MutableSet<String>,
        antonyms: MutableSet<String>,
        sentences: MutableSet<String>,
        fileUrl: String?,
        audioFileCreatedAt: LocalDateTime?
    ) : this(
        id,
        title,
        kind,
        difficulty,
        descriptions,
        synonyms,
        antonyms,
        sentences,
        emptySet(),
        fileUrl,
        audioFileCreatedAt
    )


}